llvm/lib/Target/NVPTX/NVPTXLowerArgs.cpp

```diff
@@ -734,3 +739,21 @@ bool NVPTXLowerArgs::runOnFunction(Function &F) {
 }
 
 FunctionPass *llvm::createNVPTXLowerArgsPass() { return new NVPTXLowerArgs(); }
+
+static bool copyFunctionByValArgs(Function &F) {
+  LLVM_DEBUG(dbgs() << "Creating a copy of byval args of " << F.getName()
+                    << "\n");
+  bool Changed = false;
+  for (Argument &Arg : F.args())
+    if (Arg.getType()->isPointerTy() && Arg.hasByValAttr()) {
+      copyByValParam(F, Arg);
```

Comment:
to clarify: any readonly byvals which were not formerly lowered to allocas will get lowered to allocas, but SROA will convert loads from the allocas to loads from the byval pointer?

Reply:
In a nutshell - yes. complete optimization pipeline works better than the last-ditch naive pointer use analysis + SROA pass to attempt removing the copies we made because our analysis was too simplistic.

When enabled, *all* byvals will have a copy added early. The assumption is that LLVM optimization passes will be able to remove those allocas, if possible (and that would cover the legal no-copy use cases for byval arguments), Remaining allocase are the ones that are really needed, and will not trigger additional argument copies, as copying to the alloca is done with loads only.

Reply:
Right, the byvals which get written to will require an alloca anyway, regardless of whether they are lowered early or late. Byvals which are readonly don't need an alloca, but the standard optimization pipeline will be able to figure this out and eliminate them. 

Will allocas associated with `grid_constant` parameters, which *do* get lowered in EarlyByValCopy get eliminated though? (for example, if the pointer escapes to a device function, LLVM optimization passes probably will *not* eliminate the alloca, but writes to grid constant are undefined, and the alloca _should_ be eliminated)

Reply:
Remains to be seen. We may consider taking `grid_constant` into account and skip the early copy for such byval arguments. 

I've ran a toy test with/without early copy with `bin/opt --nvptx-early-byval-copy=[0|1] -mtriple nvptx64 -mcpu=sm_70 -mattr=ptx77 -O3   -S < test/CodeGen/NVPTX/lower-byval-args.ll | bin/llc -mtriple nvptx64 -mcpu=sm_70 -mattr=ptx77`

https://gist.githubusercontent.com/Artem-B/f8af0b7f60668c4ca6b900c56d72ce23/raw/da4b7ec5c05e6914cda8b6be9738c1cbbb3ffe86/early-copy.diff

Results are... interesting, even on these toy examples. The final code is substantially smaller (16K -> 12K) and looks quite a bit better than the code w/o copies. However, `cvta.param` is gone, so that's something I'll look into in the follow-up patches.

Reply:
Sounds good.

Reply:
It looks like the missing `cvta.param` is actually an improvement.
The early copy apparently obviates the need to use `cvta.param` (e.g. see `test_select` and `test_phi` in the diff above) because LLVM managed to speculatively hoist the loads above the conditional select of pointers, and turn it into select of values. No pointers -> no problems. :-) It may not be a universal improvement, but it works nicely in this case.

I've added the check to skip early-copying `__grid_constant__` arguments, so they will not be directly affected by the early copy. I do not think I'll be able to meaningfully test it, as there's no clang-compileable CUDA code that would use the attribute, so keeping its handling unchanged makes sense.